### ISSUE TITLE
HS220 v2 hardware support?

### ISSUE THREAD

**Issue by joeio** [plasticrake/homebridge-tplink-smarthome#131]
Just received a HS220 from Amazon today. Packaging was different (smaller) than the previous two I ordered. Switch itself was identical to the other two I already have up and running with this plugin.

### Expected Behavior
Shutdown Homebridge, add new switch through the Kasa app and then restart Homebridge. Was expecting new switch to be available through HomeKit.

### Current Behavior
Device not available through HomeKit. Looking at the switch, it seems to be going through reboot loops. Even basic light control right from the switch aren't working and device shows up as offline in Kasa app. Sometimes appear to come online but sending commands fail. Uninstalling this plugin from Homebridge makes the device functional again from the Kasa app.

After several hard reset, realize that it reports as being of "Hardware Version 2.0" where my to other devices report as 1.0. Firmware and MAC address are also quite different: v1.0.3 and B0:95:75 (instead of v1.5.11 and 1C:3B:F3 for my v1.0 hardware devices).

### Versions
- **Node**: v12.18.1
- **Homebridge**: 1.1.1
- **OS**:  CentOS 7.7.1908

### Configuration
```json
        {
            "name": "TplinkSmarthome",
            "platform": "TplinkSmarthome"
        }
```

### Homebridge Log / Command Output
```
Nothing relevant shows up in the Homebridge log. Only the discovery packet being sent out and the discovery of both my v1.0 devices. 

```

**Comment by joeio** [comment]
Further troubleshooting the issue, I've tried manually setting the devices in Homebridge's config, with the same outcome. I then tried to blacklist the v2 device's MAC but that didn't help. As long as the plugin is loaded in Homebridge, I can't do anything with the switch, not even a factory reset to add it to the Kasa app again (crash loop).

With the plugin now removed from Homebridge, I did some testing with tplink-smarthome-api and I'm able to successfully retrieve its info (getSysInfo) and set its power state. Now I'm confused as to what exactly the Homebridge plugins does that completely crash the switch. Again, no log are generated that concerns this particular device, only my other two v1 switches.

It might have something to do with the plugin's discovery process but I can't find any way to disable it completely.

**Comment by plasticrake** [comment]
@joeio Great troubleshooting! I actually just completely rewrote tplink-smarthome-api (in beta) and in process of rewriting this plugin as well. It would be very helpful if you could post the results of getSysInfo here. Feel free to mask any sensitive data.

**Comment by joeio** [comment]
@plasticrake Thanks for your response!

I can now pretty much confirm that it's the discovery process that crashes the switch. I've tried narrowing the broadcast to my own subnet but that didn't help. I've also enabled debug login on my Client object and while I can see what's going on in more details, still no trace of that particular switch. It's as if the simple fact of processing a discovery packet is enough to crash it. 

Please let me know if there's any particular troubleshooting step you want me to try. 

For now, here's the getSysInfo output :
> {
  sw_ver: '1.0.3 Build 200326 Rel.082355',
  hw_ver: '2.0',
  model: 'HS220(US)',
  deviceId: '8006095114E69360F832C8235DB673F91CC6D530',
  oemId: '97D03CA037C71B6BCFAB8705E9B0C417',
  hwId: 'CA321A94521D18706FC7C34CA84F99A4',
  rssi: -30,
  latitude_i: 0,
  longitude_i: 0,
  alias: ' Hallway Light',
  mic_type: 'IOT.SMARTPLUGSWITCH',
  feature: 'TIM',
  mac: 'B0:95:75:0C:F9:95',
  updating: 0,
  led_off: 0,
  relay_state: 0,
  brightness: 100,
  on_time: 0,
  icon_hash: '',
  dev_name: 'Wi-Fi Smart Dimmer',
  active_mode: 'none',
  next_action: { type: -1 },
  preferred_state: [
    { index: 0, brightness: 100 },
    { index: 1, brightness: 75 },
    { index: 2, brightness: 50 },
    { index: 3, brightness: 25 }
  ],
  err_code: 0
}

**Comment by plasticrake** [comment]
Thanks! The discovery actually sends a longer command to include energy monitoring information and I wonder if that is causing an issue. Can you run this command as well and paste the results, in both TCP and UDP (replace DEVICEIP with your device IP):
```
tplink-smarthome-api send DEVICEIP '{"system":{"get_sysinfo":{}},"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}' 
```
And then run with `-u` for UDP:
```
tplink-smarthome-api send -u DEVICEIP '{"system":{"get_sysinfo":{}},"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}' 
```

**Comment by joeio** [comment]
Ok, both commands time out but after sending the TCP version, I can still operate and ping the switch. However, as soon as I send the UDP command, the switch hangs and then reboots.

Getting closer! 👍

**Comment by plasticrake** [comment]
Can you paste the results of the TCP version for the last command above?

Also can you try the following using UDP `-u` to see if they also crash:

```
tplink-smarthome-api send -u DEVICEIP '{"emeter":{"get_realtime":{}}}'
```

```
tplink-smarthome-api send -u DEVICEIP '{"smartlife.iot.common.emeter":{"get_realtime":{}}}'
```

**Comment by plasticrake** [comment]
Oh I misread, you never got a result from the discovery command via TCP. Please try the commands in the last post both UDP and TCP to see if they crash or if you get any results.

**Comment by joeio** [comment]
These two work fine without crashing it:

> tplink-smarthome-api send -u 10.13.0.57 '{"emeter":{"get_realtime":{}}}'
Sending to 10.13.0.57: via udp...
response:
{ emeter: { err_code: -1, err_msg: 'module not support' } }

> tplink-smarthome-api send -u 10.13.0.57 '{"smartlife.iot.common.emeter":{"get_realtime":{}}}'
Sending to 10.13.0.57: via udp...
response:
{
  'smartlife.iot.common.emeter': { err_code: -1, err_msg: 'module not support' }
}

**Comment by plasticrake** [comment]
And can you try them together without the sysinfo command?
```
tplink-smarthome-api send -u DEVICEIP '{"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}'
```

**Comment by joeio** [comment]
Still fine

> tplink-smarthome-api send -u 10.13.0.57 '{"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}'
Sending to 10.13.0.57: via udp...
response:
{
  emeter: { err_code: -1, err_msg: 'module not support' },
  'smartlife.iot.common.emeter': { err_code: -1, err_msg: 'module not support' }
}

**Comment by plasticrake** [comment]
Can you confirm that your device's firmware is up to date? This looks like a bug in the device. But either way I'll try to make the discovery configurable to exclude the energy monitoring portion. Unfortunately that would affect all devices since its broadcast and not for a particular device so I'll need to make some other workarounds.

My best guess is that there is a bug in the network handling code on the device when the UDP response is over a certain size.

**Comment by joeio** [comment]
Yes, that's something that I wonder as well because when I first tried to set it up (while the plugin was broadcasting) it told me that a firmware update was available. I accepted the prompt but it kind finished way too quickly, in my opinion. Then I started seeing all of these problems which brought me to realize that it was a different hardware revision and that the firmware version was significantly different. Maybe they started from 1.0 again with this new revision? If I try to check for new updates "manually" in the app, it says that none are available. I've tried looking on the web but couldn't find any details, release note or anything really about these things firmware releases.

I've ordered my first two switches from Amazon two weeks ago and I got v1s. Now, I've ordered another one last week and got a v2. My guess is that we'll start to see a lot more of these in the wild!

I've now ordered another one from Best Buy, at first hoping I'd get a v1 but now it could also be a way of confirming if it's a single device issue or really something different with v2s.

I'm picking up my order in 2-3 hours. I'll report back as soon as I've been able to test it all.

Thanks again for the follow up!

**Comment by joeio** [comment]
Ok, so I finally got another v2 to test and it behaves exactly the same as the first one I've got. As soon as I send the get_sysinfo command by UDP, the device disconnects and reboots.

What would be the simplest way for me to edit your code to disable any broadcasting? I'd simply add devices manually.

Thanks!

**Comment by joeio** [comment]
Nevermind, I simply edited the discoveryMsgBuf constant to remove the "emeter" part and it's working fine for my own use.

**Comment by edykim** [comment]
I've got HS220(v2) recently, and I've got the same issue. The dimmer switch is not showing on discovery phase.

I ran some commands from the thread.

```
$ npx tplink-smarthome-api send <DEVICE_IP> '{"system":{"get_sysinfo":{}},"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}' 

Sending to <DEVICE_IP>: via tcp...
TCP <DEVICE_IP>:9999 Error: TCP Timeout
    at Timeout._onTimeout (~/my/working/directory/node_modules/tplink-smarthome-api/lib/network/tcp-socket.js:62:22)
    at listOnTimeout (internal/timers.js:531:17)
    at processTimers (internal/timers.js:475:7)
Error:
Error: TCP Timeout
    at Timeout._onTimeout (~/my/working/directory/node_modules/tplink-smarthome-api/lib/network/tcp-socket.js:62:22)
    at listOnTimeout (internal/timers.js:531:17)
    at processTimers (internal/timers.js:475:7)
$ npx tplink-smarthome-api send -u <DEVICE_IP> '{"system":{"get_sysinfo":{}},"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}' 

Sending to <DEVICE_IP>: via udp...
UDP <DEVICE_IP>:9999 Error: UDP Timeout
    at Timeout._onTimeout (~/my/working/directory/node_modules/tplink-smarthome-api/lib/network/udp-socket.js:74:20)
    at listOnTimeout (internal/timers.js:531:17)
    at processTimers (internal/timers.js:475:7)
Error:
Error: UDP Timeout
    at Timeout._onTimeout (~/my/working/directory/node_modules/tplink-smarthome-api/lib/network/udp-socket.js:74:20)
    at listOnTimeout (internal/timers.js:531:17)
    at processTimers (internal/timers.js:475:7)
```

These commands are working fine.

```
$ npx tplink-smarthome-api send <DEVICE_IP> '{"system":{"get_sysinfo":{}}}'                           
Sending to <DEVICE_IP>: via tcp...
response:
{
  system: {
    get_sysinfo: {
      sw_ver: '1.0.3 Build 200326 Rel.082355',
      hw_ver: '2.0',
      model: 'HS220(US)',
      deviceId: '****',
      oemId: '****',
      hwId: '****',
      rssi: -45,
      latitude_i: 0,
      longitude_i: 0,
      alias: ' Living room light',
      mic_type: 'IOT.SMARTPLUGSWITCH',
      feature: 'TIM',
      mac: '****',
      updating: 0,
      led_off: 0,
      relay_state: 0,
      brightness: 100,
      on_time: 0,
      icon_hash: '',
      dev_name: 'Wi-Fi Smart Dimmer',
      active_mode: 'none',
      next_action: { type: -1 },
      preferred_state: [
        { index: 0, brightness: 100 },
        { index: 1, brightness: 75 },
        { index: 2, brightness: 50 },
        { index: 3, brightness: 25 }
      ],
      err_code: 0
    }
  }
}
$ npx tplink-smarthome-api send <DEVICE_IP> '{"smartlife.iot.dimmer":{"get_default_behavior":{}}}'
Sending to <DEVICE_IP>: via tcp...
response:
{
  'smartlife.iot.dimmer': {
    get_default_behavior: {
      soft_on: { mode: 'last_status' },
      hard_on: { mode: 'last_status' },
      long_press: { mode: 'instant_on_off' },
      double_click: { mode: 'gentle_on_off' },
      err_code: 0
    }
  }
}
$ npx tplink-smarthome-api send <DEVICE_IP> '{"smartlife.iot.dimmer":{"get_realtime":{}}}'        
Sending to <DEVICE_IP>: via tcp...
response:
{
  'smartlife.iot.dimmer': { get_realtime: { err_code: -2, err_msg: 'member not support' } }
}

$ npx tplink-smarthome-api send <DEVICE_IP> '{"smartlife.iot.common.emeter":{"get_default_behavior":{}}}'
Sending to <DEVICE_IP>: via tcp...
response:
{
  'smartlife.iot.common.emeter': { err_code: -1, err_msg: 'module not support' }
}
$ npx tplink-smarthome-api send <DEVICE_IP> '{"smartlife.iot.common.emeter":{"get_realtime":{}}}'
Sending to <DEVICE_IP>: via tcp...
response:
{
  'smartlife.iot.common.emeter': { err_code: -1, err_msg: 'module not support' }
}
```

**Comment by edykim** [comment]
HS220 v2 timed out while discover with current `discoveryMsgBuf`. Also, I can confirm that HS220 is kinda broken after discovery so I need to restart the dimmer. 

```
$ npx tplink-smarthome-api send <DEVICE_IP> '{"system":{"get_sysinfo":{}},"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}'         
Sending to <DEVICE_IP>: via tcp...
TCP <DEVICE_IP>:9999 Error: TCP Timeout
```

The device respond properly when I omit `get_sysinfo`.

```
$ npx tplink-smarthome-api send <DEVICE_IP> '{"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}' 
Sending to <DEVICE_IP>: via tcp...
response:
{
  emeter: { err_code: -1, err_msg: 'module not support' },
  'smartlife.iot.common.emeter': { err_code: -1, err_msg: 'module not support' }
}
```

**Comment by patelhiren** [comment]
I will add myself as another user who just received **hw_ver 2.0 (HS220(US))** switches from Amazon. Exact same behavior. Enabling the plugin in homebridge breaks the access to the switch from Kasa app and it seems to be cauing the switch to go into reboot loop.

Haven't spent much time troubleshooting other than running the `get_sysinfo` command. Output reports as below.

```
sw_ver: '1.0.3 Build 200326 Rel.082355',
hw_ver: '2.0',
model: 'HS220(US)',
```

**Comment by joeio** [comment]
If you guys don't need the energy monitoring portion of the api, you can keep using the plugin by manually editing the file

`homebridge/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib/client.js`

At line 19, replace: 
`const discoveryMsgBuf = encrypt(
  '{"system":{"get_sysinfo":{}},"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}'
);`

with

`const discoveryMsgBuf = encrypt('{"system":{"get_sysinfo":{}}}');`

Keep in mind that this will break as soon as the plugin is updated or removed and reinstalled in Homebridge. 

For my part, I've been running with this edit for the past two weeks without any issues.

**Comment by patelhiren** [comment]
Thanks made that change locally and I can confirm it works very well.

> If you guys don't need the energy monitoring portion of the api, you can keep using the plugin by manually editing the file
> 
> `homebridge/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib/client.js`
> 

> At line 19, replace:
> `const discoveryMsgBuf = encrypt( '{"system":{"get_sysinfo":{}},"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}' );`
> 
> with
> 
> `const discoveryMsgBuf = encrypt('{"system":{"get_sysinfo":{}}}');`
> 
> Keep in mind that this will break as soon as the plugin is updated or removed and reinstalled in Homebridge.
> 
> For my part, I've been running with this edit for the past two weeks without any issues.

**Comment by KeshavMadhavan** [comment]
Hey quick question. What file exactly are you manually editing? I can't find a node_modules folder in my .homebridge folder on my raspberry pi. Thanks

**Comment by KeshavMadhavan** [comment]
Hello so I tried going to this path. I got all the way to the last folder but stopped because I don't see a client.js file in the lib folder. Any suggestions? Thanks

> If you guys don't need the energy monitoring portion of the api, you can keep using the plugin by manually editing the file
> 
> `homebridge/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib/client.js`
> 
> At line 19, replace:
> `const discoveryMsgBuf = encrypt( '{"system":{"get_sysinfo":{}},"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}' );`
> 
> with
> 
> `const discoveryMsgBuf = encrypt('{"system":{"get_sysinfo":{}}}');`
> 
> Keep in mind that this will break as soon as the plugin is updated or removed and reinstalled in Homebridge.
> 
> For my part, I've been running with this edit for the past two weeks without any issues.

**Comment by KeshavMadhavan** [comment]
> If you guys don't need the energy monitoring portion of the api, you can keep using the plugin by manually editing the file
> 
> `homebridge/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib/client.js`
> 
> At line 19, replace:
> `const discoveryMsgBuf = encrypt( '{"system":{"get_sysinfo":{}},"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}' );`
> 
> with
> 
> `const discoveryMsgBuf = encrypt('{"system":{"get_sysinfo":{}}}');`
> 
> Keep in mind that this will break as soon as the plugin is updated or removed and reinstalled in Homebridge.
> 
> For my part, I've been running with this edit for the past two weeks without any issues.

I made this change and still can't get it working. Anything else I have to do other than this?

**Comment by budliter** [comment]
So - I am having the same problem.  I tried the fix listed above - but I can't figure out how to edit the file directly in either raspberry pi or through the homebridge interface.  In file manager in the raspberry pi, I don't have write permissions since it belongs to the homebridge user and I can figure out how to edit directly in the homebridge terminal. 

What is the best way to edit the file?

Thank you

**Comment by elementdc5** [comment]
I am experiencing the same problem

**Comment by elementdc5** [comment]
> So - I am having the same problem. I tried the fix listed above - but I can't figure out how to edit the file directly in either raspberry pi or through the homebridge interface. In file manager in the raspberry pi, I don't have write permissions since it belongs to the homebridge user and I can figure out how to edit directly in the homebridge terminal.
> 
> What is the best way to edit the file?
> 
> Thank you

Any luck on yours?

**Comment by songo-github** [comment]
> If you guys don't need the energy monitoring portion of the api, you can keep using the plugin by manually editing the file
> 
> `homebridge/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib/client.js`
> 
> At line 19, replace:
> `const discoveryMsgBuf = encrypt( '{"system":{"get_sysinfo":{}},"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}' );`
> 
> with
> 
> `const discoveryMsgBuf = encrypt('{"system":{"get_sysinfo":{}}}');`
> 
> Keep in mind that this will break as soon as the plugin is updated or removed and reinstalled in Homebridge.
> 
> For my part, I've been running with this edit for the past two weeks without any issues.

can confirm this works to fix the error when using HS300. Only seeing this when a device is present that has power monitoring. Second location with only plugs and switches with no monitoring has no issues

**Comment by budliter** [comment]
> > So - I am having the same problem. I tried the fix listed above - but I can't figure out how to edit the file directly in either raspberry pi or through the homebridge interface. In file manager in the raspberry pi, I don't have write permissions since it belongs to the homebridge user and I can figure out how to edit directly in the homebridge terminal.
> > What is the best way to edit the file?
> > Thank you
> 
> Any luck on yours?

No I gave up trying for now - I can't figure out how to get around the permissions to manually edit.  So, my broader question is 1) Do you manually edit in the homebridge editor interface or 2) Do you do it logged into the raspberry pi.  For #1 - I can't figure out how to open something to edit the file in the command interface.  For #2 - I do not have permission.

**Comment by bearattack11** [comment]
> If you guys don't need the energy monitoring portion of the api, you can keep using the plugin by manually editing the file
> 
> `homebridge/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib/client.js`
> 
> At line 19, replace:
> `const discoveryMsgBuf = encrypt( '{"system":{"get_sysinfo":{}},"emeter":{"get_realtime":{}},"smartlife.iot.common.emeter":{"get_realtime":{}}}' );`
> 
> with
> 
> `const discoveryMsgBuf = encrypt('{"system":{"get_sysinfo":{}}}');`
> 
> Keep in mind that this will break as soon as the plugin is updated or removed and reinstalled in Homebridge.
> 
> For my part, I've been running with this edit for the past two weeks without any issues.

Thanks, can confirm this fixed my issue with the device appearing. However, my HS220 shows up as a static switch without the dimmer function. Is there anyway to add dimming/brightness control?

**Comment by joeio** [comment]
> Thanks, can confirm this fixed my issue with the device appearing. However, my HS220 shows up as a static switch without the dimmer function. Is there anyway to add dimming/brightness control?

This "fix" has nothing to do with that so it must be something else you changed. Dimming and brightness control work fine for me.

**Comment by elementdc5** [comment]
@joeio My directory goes like this - /docker/homebridge/node_modules/homebridge-dummy  
I can't find any node module for tplink.  Any ideas?  I'm running homebridge in docker on a synology nas

**Comment by joeio** [comment]
> @joeio My directory goes like this - /docker/homebridge/node_modules/homebridge-dummy
> I can't find any node module for tplink. Any ideas? I'm running homebridge in docker on a synology nas

It can be very different between setups but if it's the correct starting directory, you should have something like this (_homebridge-dummy_ is just another module)
`/docker/homebridge/node_modules/homebridge-tplink-smarthome/...`

**Comment by elementdc5** [comment]
> 
> 
> > @joeio My directory goes like this - /docker/homebridge/node_modules/homebridge-dummy
> > I can't find any node module for tplink. Any ideas? I'm running homebridge in docker on a synology nas
> 
> It can be very different between setups but if it's the correct starting directory, you should have something like this:
> `/docker/homebridge/node_modules/homebridge-tplink-smarthome/...`
> 
> (_homebridge-dummy_ is just another module unrelated to this issue)

Yeah I can't figure out where its at.  homebridge-dummy is the only folder I have under node_modules

**Comment by songo-github** [comment]
> > > So - I am having the same problem. I tried the fix listed above - but I can't figure out how to edit the file directly in either raspberry pi or through the homebridge interface. In file manager in the raspberry pi, I don't have write permissions since it belongs to the homebridge user and I can figure out how to edit directly in the homebridge terminal.
> > > What is the best way to edit the file?
> > > Thank you
> > 
> > 
> > Any luck on yours?
> 
> No I gave up trying for now - I can't figure out how to get around the permissions to manually edit. So, my broader question is 1) Do you manually edit in the homebridge editor interface or 2) Do you do it logged into the raspberry pi. For #1 - I can't figure out how to open something to edit the file in the command interface. For #2 - I do not have permission.
Are you able to open a terminal window on the pi and type sudo -i ?

**Comment by budliter** [comment]
Ok - so I figured out how to edit the code per the information above, but haven't tested if the dimmer works now.  Here is how to edit the file:

- You need to update your file manager so you log in as the root user.  Note - this approach is a bit risky as you can now edit all system files, so you may want to change back when done.  1)  Go to menu>preferences>main menu editor>click on accessories then "new item".  2) In the launcher properties - Name:  Root File Manager | Command: sudo pcmanfm | Comment: Open the file manager as root;
- This will create a new icon in the accessory drop down for a file manager that is now running as the root user - so all files are editable
- Run the new icon you created
- Click view and check "show hidden"
- Then find the file - for me it is in /usr/lib/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib
-Right click on the client.js file and edit in Geany...then save

**Comment by plasticrake** [comment]
I have a pre-release version of this plugin that is a complete rewrite from javascript to typescript. And one of the changes includes a fix for this issue. You can install it using the next tag like: `npm -g install homebridge-tplink-smarthome@next`

**Comment by elementdc5** [comment]
> npm -g install homebridge-tplink-smarthome@next

Working great!  Thank you

**Comment by plasticrake** [comment]
I've released v5.2.0 which should fix this issue. @joeio please let me know if this resolved the issue for you!

**Comment by plasticrake** [comment]
I haven't heard any feedback so I'm going to assume this is resolved. Closing
